HeuristicNew.java

```diff
@@ -0,0 +1,256 @@
+        int count = 0;
+        int numRows = field.length;
+        int numCols = field[0].length;
+        for (int j = 0; j < numCols; j++) {
+            for (int i = 0; i < numRows - 1; i++) {
+                if (field[i][j] == 0 && field[i+1][j] != 0
+                        || field[i][j] != 0 && field[i+1][j] == 0){
+                    count++;
+                }
+            }
+        }
+        return count;
+    }
+
+    // Returns array of column heights
+    private static int[] colHeights(int[][] field) {
+        int[] colHeights = new int[field[0].length];
+
+        // System.out.print("Col heights: ");
+
+        for (int col = 0; col < field[0].length; col++) {
+            int row = field.length - 1;
+
+            while (row >= 0 && field[row][col] == 0) {
+                row--;
+            }
+            colHeights[col] = row + 1;
+            // System.out.print(colHeights[col] + " ");
+        }
+        // System.out.println();
+
+        return colHeights;
+    }
+
+    // Feature 5
+    // Returns weighted number of holes
+    private static double wNumHoles(int[][] field, int[] colHeights) {
+        int numHoles = 0;
+
+        for (int col = 0; col < colHeights.length; col++) {
+            for (int row = colHeights[col] - 1; row >= 0; row--) {
+                if (field[row][col] == 0) {
+                    numHoles++;
+                }
+            }
+        }
+
+        // System.out.println("Number of holes = " + numHoles);
+
+        return getNextWeight() * numHoles;
+    }
+
+    // Feature 6
+    private static double weightedWellSums(int[][] field) {
+        return getNextWeight() * numColTransitions(field);
+    }
+
+    // A well is a succession of empty cells such that their left cells and right cells are both filled.
+    // Example:
+    // 1101100010
+    // 1101101010
+    // 1101101000
+    // returns 8
+    private static int wellSums(int[][] field) {
+        int count = 0;
+        int numRows = field.length;
+        int numCols = field[0].length;
+        for (int i = 0; i < numRows; i++) {
+            for (int j = 0; j < numCols; j++) {
+                if (field[i][j] == 0){ // if cell is empty
+                        if (j > 0 && j < numCols - 1 && field[i][j - 1] != 0 && field[i][j + 1] != 0) {
+                            count += exploreWell(i, j, field);
+                        } else if (j == 0 && field[i][j+1] !=0) {
+                            count += exploreWellLeftSide(i, field);
+                        } else if (j == numCols - 1 && field[i][j-1] != 0){
+                            count += exploreWellRightSide(i, field);
+                        }
+                }
+
+            }
+        }
+        //System.out.println("Overall Well sum: " + count);
+        return count;
+    }
+
+    // explore up the rows within the column
+    private static int exploreWell(int startRow, int startColumn, int[][] field) {
```

Comment:
actually if you don't make this a separate method, you can just do

```
for i=0:rowLength
    for j=0:colLength - 1
        if field[i][j] == 0
            while field[i][j]==0 // to find entire well
                j++
            count++
```

+ all the edge cases you have identified? Right?

Reply:
while field[i][j]==0 doesn't necessarily mean that you're explore up a well cos it's an empty cell, you might be exploring a hole. and i'm not sure how your suggested algo covers this case. 1's are occupied, 0 are empty
110
001
101
101 

your suggestion will sum all 3 empty cells when it is supposed to count 2 cells.